runners/flink/src/main/java/org/apache/beam/runners/flink/translation/functions/ReferenceCountingFlinkExecutableStageContextFactory.java

```diff
@@ -104,9 +108,21 @@ private void scheduleRelease(JobInfo jobInfo) {
     WrappedContext wrapper = getCache().get(jobInfo.jobId());
     Preconditions.checkState(
         wrapper != null, "Releasing context for unknown job: " + jobInfo.jobId());
-    // Do not release this asynchronously, as the releasing could fail due to the classloader not being
-    // available anymore after the tasks have been removed from the execution engine.
-    release(wrapper);
+
+    PipelineOptions pipelineOptions =
+        PipelineOptionsTranslation.fromProto(jobInfo.pipelineOptions());
+    int environmentCacheTTLMillis =
+        pipelineOptions.as(PortablePipelineOptions.class).getEnvironmentCacheMillis();
+    if (environmentCacheTTLMillis > 0) {
+      // Schedule task to clean the container later.
+      // Ensure that this class is loaded in the parent Flink classloader.
+      getExecutor()
+          .schedule(() -> release(wrapper), environmentCacheTTLMillis, TimeUnit.MILLISECONDS);
```

Comment:
How do we ensure that the class is loaded? We could perform a release before we schedule the release.

Reply:
Class is loaded in the current class loader. however I am not sure how to check if the class is loaded in flink class loader without making things too complicated.

> We could perform a release before we schedule the release.

But we want to differ the release as we want to cache.

Reply:
It isn't possible to defer this when the class is loaded by the user class loader. So we would probably need to compare the class loader that loaded the flink runtime with that of the beam class?

Reply:
Something like this (disclaimer: not tested):

```
if (environmentCacheTTLMillis > 0 && this.getClass().getClassLoader() == ExecutionEnvironment.class.getClassLoader()) 
```
For execution in the job server, class loader will be same (applies for Jenkins). On the remote Flink cluster (by default), the user class loader will be different and will be removed.

Reply:
@tweise Makes sense. I will add that check. In addition of falling back to immediate release, I will also log a warning.

Reply:
@mxm The class is first loaded when we create the environment. So we can be assured that the class is loaded before release. 
Also, we require classes to be loaded on parent classloader for async container destruction. This inherently mean that once the class is loaded in parent class loader, its not going to be unloaded in this scenario. With the additional check mentioned, we will enforce this requirement.